ecllibrary/std/DataPatterns/Benford.ecl

```diff
@@ -0,0 +1,314 @@
+ *                          OPTIONAL, defaults to an empty string
+ * @param   digit           The 1-based digit within the number to examine; the
+ *                          first significant digit is '1' and it only increases;
+ *                          OPTIONAL, defaults to 1, meaning the most-significant
+ *                          non-zero digit
+ * @param   sampleSize      A positive integer representing a percentage of
+ *                          inFile to examine, which is useful when analyzing a
+ *                          very large dataset and only an estimated data
+ *                          analysis is sufficient; valid range for this
+ *                          argument is 1-100; values outside of this range
+ *                          will be clamped; OPTIONAL, defaults to 100 (which
+ *                          indicates that all rows in the dataset will be used)
+ *
+ * @return  A new dataset with the following record structure:
+ *
+ *      RECORD
+ *          STRING      attribute;   // Name of data attribute examined
+ *          DECIMAL4_1  zero;        // Percentage of rows with digit of '0'
+ *          DECIMAL4_1  one;         // Percentage of rows with digit of '1'
+ *          DECIMAL4_1  two;         // Percentage of rows with digit of '2'
+ *          DECIMAL4_1  three;       // Percentage of rows with digit of '3'
+ *          DECIMAL4_1  four;        // Percentage of rows with digit of '4'
+ *          DECIMAL4_1  five;        // Percentage of rows with digit of '5'
+ *          DECIMAL4_1  six;         // Percentage of rows with digit of '6'
+ *          DECIMAL4_1  seven;       // Percentage of rows with digit of '7'
+ *          DECIMAL4_1  eight;       // Percentage of rows with digit of '8'
+ *          DECIMAL4_1  nine;        // Percentage of rows with digit of '9'
+ *          DECIMAL7_3  chi_squared; // Chi-squared "fitness test" result
+ *          UNSIGNED8   num_values;  // Number of rows with non-zero values for this attribute
+ *      END;
+ *
+ * The named digit fields (e.g. "zero" and "one" and so on) represent the
+ * digit found in the 'digit' position of the associated attribute.  The values
+ * that appear there are percentages.  num_values shows the number of
+ * non-zero values processed, and chi_squared shows the result of applying
+ * that test using the observed vs expected distribution values.
+ *
+ * The first row of the results will show the expected values for the named
+ * digits, with "-- EXPECTED DIGIT n --" showing as the attribute name.'n' will
+ * be replaced with the value of 'digit' which indicates which digit position
+ * was examined.
+ *
+ * Note that when viewing the results for the mosts significant digit (digit = 1),
+ * the 'zero' field will show a -1 value, indicating that it was ignored.
+ */
+EXPORT Benford(inFile, fieldListStr = '\'\'', digit = 1, sampleSize = 100) := FUNCTIONMACRO
+
+    #UNIQUENAME(minDigit);
```

Comment:
You seem to be going through a lot of effort to make sure that code cannot be shared between uses here.

I suspect Benford should be a MODULE rather than a FUNCTIONMACRO

Reply:
Indeed, I do this (#UNIQUENAME et al) on purpose.  I've run into many cases where calling the same function macro multiple times from the same scope creates problems due to symbol reuse.  This particular function is a good example of a function that could be called multiple times with slightly different arguments in the same scope.  Using #UNIQUENAME with LOCAL for an ECL attribute may be overkill, though, I admit.

I'm curious about using a MODULE in place of a FUNCTIONMACRO:  How would the code perform the macro-level introspection?